### PR TITLE
stabilize macro_lifetime_matcher

### PR DESCRIPTION
**Pull request by durka**
This stabilizes `:lifetime` which has completed FCP in #34303.

### CODE REVIEW

File: src/librustc/lib.rs [hunk 1]
```diff
@@ -50,7 +50,6 @@
 #![feature(from_ref)]
 #![feature(fs_read_write)]
 #![cfg_attr(windows, feature(libc))]
-#![feature(macro_lifetime_matcher)]
```
Comment (kennytm) [inline src/librustc/lib.rs]: You still need `#![cfg_attr(stage0, feature(macro_lifetime_matcher))]` for these.

**Comment by nikomatsakis** [comment]
cc @rust-lang/docs -- what docs need updating for this? It's the stabilization of the lifetime matcher in macros.

**Comment by durka** [comment]
I sent a PR to the reference (and the book, but only the first edition has
this level of detail on macro_rules and I guess that's frozen now).

**Comment by GuillaumeGomez** [comment]
In the rust book maybe?

**Comment by durka** [comment]
Yeah, probably, since I haven't gotten around to adding the `cfg(stage0)` stuff yet. But I hope that the decision to merge `:ident` and `:lifetime` is **not** made, because it wouldn't be possible for `macro_rules` to tell the difference in that case. (I guess in a proc macro you can check the name, but that seems fragile.) Moreover, this code would change meaning:

```rust
macro_rules! m {
    (T: $i:ident) => { true };
    (T: $t:tt) => { false }
}

m!(T: 'static)
```

**Comment by durka** [comment]
@petrochenkov that other thread is about the proc macro API. If you still want to merge ident and lifetime for macro_rules, can you explain the breaking change motivation?

**Comment by petrochenkov** [comment]
@durka 
After some reflecting, I agree, the motivation for merging them in declarative macros is weak compared to drawbacks.
Even from teaching point of view, I don't think people think of lifetimes and normal identifiers as "same things", so it'd probably mostly result in extra WTFs-per-second.

>making them indistinguishable from idents

(I expected `lifetime` to be a *subset* of `ident` (https://github.com/rust-lang/rust/pull/50473#issuecomment-387205027), so they would be distinguishable.)

**Comment by durka** [comment]
> subset

Oh I see, maybe I misunderstood. I thought you were proposing to merge them so :lifetime wouldn't be added, and `'foo` would match :ident. That'd be a problem because you couldn't have a macro rules that matches lifetimes-but-not-idents.

BTW the idea of `'$lt:ident` was also discussed back when :lifetime was proposed, but it didn't seem to get traction.

**Comment by petrochenkov** [comment]
I don't think so.
@bors r+

**Pull request merged**